WalletWasabi/Wallets/UnconfirmedTransactionChainProvider.cs

```diff
@@ -0,0 +1,136 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Logging;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class UnconfirmedTransactionChainProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumRequestsInParallel = 3;
+
+	public UnconfirmedTransactionChainProvider(WasabiHttpClientFactory httpClientFactory)
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler<EventArgs>? RequestedUnconfirmedChainArrived;
+
+	public ConcurrentDictionary<uint256, List<UnconfirmedTransactionChainItem>> UnconfirmedChainCache { get; } = new();
+	public ConcurrentQueue<uint256> Queue { get; } = new();
+	private SemaphoreSlim Semaphore { get; } = new(initialCount: 0, maxCount: MaximumRequestsInParallel);
+
+	private IHttpClient HttpClient { get; }
+
+	private async Task FetchUnconfirmedTransactionChainAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		const int MaxAttempts = 3;
+
+		for (int i = 0; i < MaxAttempts; i++)
+		{
+			try
+			{
+				using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(60));
+				using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+				var response = await HttpClient.SendAsync(
+					HttpMethod.Get,
+					$"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/unconfirmed-transaction-chain?transactionId={txid}",
+					null,
+					linkedCts.Token).ConfigureAwait(false);
+
+				if (response.StatusCode != HttpStatusCode.OK)
+				{
+					await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				var unconfirmedChain = await response.Content.ReadAsJsonAsync<UnconfirmedTransactionChainItem[]>().ConfigureAwait(false);
+
+				if (!UnconfirmedChainCache.TryAdd(txid, unconfirmedChain.ToList()))
+				{
+					throw new InvalidOperationException($"Failed to cache unconfirmed tx chain for {txid}");
+				}
+
+				RequestedUnconfirmedChainArrived?.Invoke(this, EventArgs.Empty);
+			}
+			catch (Exception ex)
+			{
+				if (cancellationToken.IsCancellationRequested)
+				{
+					return;
+				}
+
+				Logger.LogWarning($"Attempt: {i}. Failed to fetch transaction fee. {ex}");
+			}
+		}
+	}
+
+	public void BeginRequestUnconfirmedChain(SmartTransaction tx)
+	{
+		if (!tx.Confirmed && tx.ForeignInputs.Count != 0 && !Queue.Any(x => x == tx.GetHash()) && !UnconfirmedChainCache.ContainsKey(tx.GetHash()))
```

Comment:
`tx.ForeignInputs.Count` What if only the parent tx or in the chain has ForeignInputs?

Reply:
> What if only the parent tx has ForeignInputs

I don't get this. I don't think it's possible. 
In order to make TX1 which only uses my inputs (no foreign inputs), I needed to receive some coins from an outside source, which HAS Foreign input and we request that tx from the backend.

> in the chain has ForeignInputs?

So, if I have TX1 which has no foreign inputs (payed for something in a random shop) and the shopkeeper builds TX2 on top of my TX1, then 
1. my wallet is not aware of this, 2. It shouldn't be aware of this, 3. No need to request anything from the backend.

I might be mistaken, i'm always happy to be corrected with examples.

Reply:
> I don't think it's possible.

I was thinking of receiving `tx1` and then the user builds a transaction on top of that `tx2`. Still both unconfirmed. 

`tx2` using unconfirmed coins from `tx1`. `tx2` does not have foreign input. However, `tx1` will have a role as well in the effective feeRafe calculation.

Reply:
Well in your example, if we want `tx2` to have an effective fee rate, then we can and should calculate it on the client side.

Right now, `tx1` will be requested from the backend and the client can calculate `tx2` fee on its own. So, after the request arrives, the client will know the fee and size of both transaction, therefore the client can calculate the effective feeRate of `tx2` without requesting `tx2` from the backend.

This is not yet implemented and this PR doesn't aim to solve this.